### DISCUSSION TITLE
请教，为何我的batchmds:runv没有执行到呢？

### DISCUSSION THREAD

**Discussion by DanielKui** [xmake-io/xmake#3914] [Q&A]
以下lua脚本生成makefile后如下：

![image](https://github.com/xmake-io/xmake/assets/578118/6ace720d-23b2-4f79-935d-8bbd0b6e8f63)

我是想实现cd ./bin子目录，然后执行ln -sf libprotobuf-lite.so.23 libprotobuf-lite.so
但以上makefile似乎没有执行到。
我改为[ln](batchcmds:ln)也不行，也没执行。 不知是什么原因呢？

---		batchcmds:ln("libprotobuf-lite.so.23","libprotobuf-lite.so",{"-sf"})

includes ("../xmake.lua")
includes ("../Tools/LogManager/xmake.lua")
includes ("HXQuoteDemoFiles.lua")

add_rules("mode.debug", "mode.release")

rule("copylib")
    after_buildcmd(function (target, batchcmds, opt)
		batchcmds:cp("../ThirdParty/lib64_2022/libprotobuf-lite.so.23", target:targetdir())
		batchcmds:cp("../ThirdParty/lib64_2022/libACE.so.6.5.9", target:targetdir())
		
        	batchcmds:cd(target:targetdir())
		batchcmds:runv("ln", {"-sf","libprotobuf-lite.so.23","libprotobuf-lite.so"})
		batchcmds:cd("..")

		batchcmds:cd(target:targetdir())
		batchcmds:runv("ln", {"-sf", "libACE.so.6.5.9","libACE.so"})
		batchcmds:cd("..")

    end)

	
target ("HXQuoteApi")
	set_optimize("faster")
	set_languages(cppVersion)
	add_ldflags(linkOptions)
    set_kind ("shared")
	set_targetdir(tarDir)
	add_ldflags("-Wl,-rpath="..tarDir)
	
	add_rules("copylib")
	
    add_defines (ApiDefine)
	add_includedirs(ApiSrcDir)
	add_files(ApiCC)
    add_linkdirs(ApiLibDir)
    add_links(ApiLib)
    
	before_link(function (target)
		os.cp("../Tools/LogManager/bin/libLogManager_x64.so", target:targetdir())
		os.cp("../ThirdParty/lib64_2022/libfmt.so*", target:targetdir())
		os.cp("../ThirdParty/lib64_2022/libquill.so*", target:targetdir())
		os.cp("../ThirdParty/lib64_2022/libprotobuf-lite.so.23", target:targetdir())
		os.cp("../ThirdParty/lib64_2022/libACE.so.6.5.9", target:targetdir())
		
		os.cd(target:targetdir())
		os.run("ln -sf libprotobuf-lite.so.23 libprotobuf-lite.so")
		os.cd("..")
		
		os.cd(target:targetdir())
		os.run("ln -sf libACE.so.6.5.9 libACE.so")
		os.cd("..")
    end)
	
	if is_mode("debug") then
		add_defines("DEBUG")
		set_basename("HXQuoteApid_x64")
		set_symbols("debug")
		set_optimize("none")
		add_cxflags(debugOptions)
	end
	if is_mode("release") then
		add_defines("NDEBUG")
		set_basename("HXQuoteApi_x64")
		set_optimize("faster")
		add_cxflags(releaseOptions)
	end
	
	on_clean(function (target)
        -- Delete only target files
        os.rm(target:targetfile())
    end)

target ("HXQuoteDemo")
	set_kind ("binary")
	set_languages(cppVersion)
	add_ldflags(linkOptions)    
	set_targetdir(tarDir)
    add_defines (DemoDefine)
	
	add_includedirs(DemoSrcDir)
	add_files(DemoCC)
    add_linkdirs(DemoLibDir)
    add_links(DemoLib)
	add_deps("HXQuoteApi")
	on_clean(function (target)
        -- Delete only target files
        os.rm(target:targetfile())
    end)

	if is_mode("debug") then
		add_defines("DEBUG")
		set_basename("HXQuoteDemod_x64")
		set_symbols("debug")
		set_optimize("none")
		add_cxflags(debugOptions)
		add_links("HXQuoteApid_x64")
	end
	if is_mode("release") then
		add_defines("NDEBUG")
		set_basename("HXQuoteDemo_x64")
		set_optimize("faster")
		add_links("HXQuoteApi_x64")
		add_cxflags(releaseOptions)
	end

盼复，非常感谢

batchcmds:cp是可以执行到的

**Comment by waruqi** [comment]
> batchcmds:ln("libprotobuf-lite.so.23","libprotobuf-lite.so",{"-sf"})

makefile generator 目前没对 ln 做支持。会直接忽略。。

> batchcmds:runv("ln", {"-sf","libprotobuf-lite.so.23","libprotobuf-lite.so"})

runv 也是，目前也没支持。。

你可以提 pr 过来改进 makefiler generator 。。或者直接执行 xmake 构建，xmake 的重点在直接 build ，对 generator 生成 不会支持的像构建那么完善，尤其是 makefile 这种，对于 xmake 来说，都是可选特性，不会做重点处理。。

**Comment by DanielKui** [comment]
感谢回复，是否有能实现类似功能的语法呢？execv也不行是吗？

**Reply by waruqi** [reply]
能实现，但暂时没时间，你可以提 feature request 然后耐心等，也可以直接提 pr 帮忙实现掉
